cores/esp8266/base64.cpp

```diff
@@ -39,33 +38,21 @@ String base64::encode(const uint8_t * data, size_t length, bool doNewLines) {
     // base64 needs more size then the source data, use cencode.h macros
     size_t size = ((doNewLines ? base64_encode_expected_len(length)
                                : base64_encode_expected_len_nonewlines(length)) + 1);
-    char * buffer = (char *) malloc(size);
-    if(buffer) {
-        base64_encodestate _state;
-        if(doNewLines)
-        {
-            base64_init_encodestate(&_state);
-        }
-        else 
-        {
-            base64_init_encodestate_nonewlines(&_state);
-        }
-        int len = base64_encode_block((const char *) &data[0], length, &buffer[0], &_state);
-        len = base64_encode_blockend((buffer + len), &_state);
+    String base64;
+    if (!base64.reserve(size)) {
+        return String("-FAIL-");

```

Comment:
If here you do:
```cpp
base64 = "-FAIL-";
return base64;
```

and then you change the last return statement to:
```cpp
return base64;
```

then the compiler will be enabled for NRVO, aka copy elision, so no copies at all.

Reply:
yes, but I do have to call c_str() to do a deep copy because otherwise len() is set incorrectly to the initial value of 0 (it is not modified by the writeable begin()) . 

this is the part where we'd have to extend the public api of WString or add friend access.

Reply:
changed by using friend access

Reply:
friend access is a step in the wrong direction: it pollutes String with knowledge about base64, which has nothing to do with it. Please revert.
I don't follow your explanation about calling c_str(), wrong len, and deep copy. Please provide further details.

Reply:
I've removed the friend access. the issue is that String() is handling a string with 3 attributes: a buffer, a  capacity and a length. the capacity is the size of the memory allocation of the buffer, and the length ist he actual length of the current string stored (e.g. this is an optimisation instead of having to call strlen(buffer) all the time. 

What I did was to increase capacity of the buffer to the desired size, and then pass the buffer to the libb64 functions as the target output buffer. This works fine except that nothing ever modifies the "length" attribute of the string, and for that I need to be able to call setLen() (which is protected, hence friend). 

The only other option is to "append" a char buffer, which is less efficient - it calls strlen() on the buffer.